src/storage/tde.c

```diff
@@ -0,0 +1,1707 @@
+   * otherwise it will try to read it from cache using root class OID. which actually has its own heap file and its own
+   * heap file type.
+   */
+  error_code = heap_scancache_start_modify (thread_p, &scan_cache, &tde_Keyinfo_hfid, NULL, SINGLE_ROW_UPDATE, NULL);
+  if (error_code != NO_ERROR)
+    {
+      ASSERT_ERROR ();
+      return error_code;
+    }
+
+  /* hack the class to avoid heap_scancache_check_with_hfid. */
+  scan_cache.node.class_oid = *oid_Root_class_oid;
+  heap_create_update_context (&update_context, &tde_Keyinfo_hfid, &tde_Keyinfo_oid, oid_Root_class_oid, &recdes,
+			      &scan_cache, UPDATE_INPLACE_CURRENT_MVCCID);
+  error_code = heap_update_logical (thread_p, &update_context);
+  if (error_code != NO_ERROR)
+    {
+      ASSERT_ERROR ();
+    }
+  heap_scancache_end (thread_p, &scan_cache);
+  return error_code;
+}
+
+/*
+ * tde_chnage_mk () - Change the master key set on the database
+ *
+ * return             : Error code
+ * thread_p (in)      : Thread entry
+ * mk_index (in)      : The index in the master key file
+ * master_key (in)    : The master key
+ * created_time (in)  : The creation time of the master key
+ */
+int
+tde_change_mk (THREAD_ENTRY * thread_p, const int mk_index, const unsigned char *master_key, const time_t created_time)
+{
+  TDE_KEYINFO keyinfo;
+  TDE_DATA_KEY_SET dks;
+  int err = NO_ERROR;
+
+  if (!tde_Cipher.is_loaded)
+    {
+      err = ER_TDE_CIPHER_IS_NOT_LOADED;
+      er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_TDE_CIPHER_IS_NOT_LOADED, 0);
+      return err;
+    }
+
+  /* generate keyinfo from tde_Cipher and update heap (on Disk) */
+  err = tde_generate_keyinfo (&keyinfo, mk_index, master_key, created_time, &tde_Cipher.data_keys);
+  if (err != NO_ERROR)
+    {
+      return err;
+    }
+
+  err = tde_update_keyinfo (thread_p, &keyinfo);
+  if (err != NO_ERROR)
+    {
+      return err;
+    }
+
+  /* heap_flush() is mandatory. Without this, it cannot be guaranteed
+   * that the master key corresponding key info in heap exists in _keys file.
+   * By calling heap_flush at end of changing key, DBA can remove the previous key after it.
+   */
+  heap_flush (thread_p, &tde_Keyinfo_oid);
+
+  return err;
+}
+
+/*
+ * tde_load_mk () - Read a master key specified by keyinfo from the master key file and validate it
+ *
+ * return             : Error code
+ * vdes (in)          : Key file descriptor
+ * keyinfo (in)       : Key info which contains the master key information to load
+ * master_key (out)   : Master key
+ */
+int
+tde_load_mk (int vdes, const TDE_KEYINFO * keyinfo, unsigned char *master_key)
+{
+  int err = NO_ERROR;
+  unsigned char mk[TDE_MASTER_KEY_LENGTH];
+  time_t created_time;
+
+  assert (keyinfo->mk_index >= 0);
+
+  err = tde_find_mk (vdes, keyinfo->mk_index, mk, &created_time);
+  if (err != NO_ERROR)
+    {
+      return err;
+    }
+
+  /* MK has found */
+
+  if (!(tde_validate_mk (mk, keyinfo->mk_hash) && created_time == keyinfo->created_time))
```

Comment:
"&& created_time == keyinfo->created_time" -> "|| created_time != keyinfo->created_time". isn't this right?

Reply:
I think it's right. I suspect you missed the parentheses which embrace `tde_validate_mk()` and `created_time == keyinfo->created_time`